kori/app/dao/product.py

```diff
@@ -0,0 +1,152 @@
+            valid_from=datetime.now(),
+            valid_to=datetime(9999, 12, 31, 23, 59, 59, 999999),
+        )
+
+        session.add(new_product_version_db)
+        session.commit()
+    except IntegrityError:
+        raise NotFoundException(message="Organisation not found.")
+
+    return ProductSchema(
+        product_id=new_product_db.id,
+        reorder_level=new_product_db.reorder_level,
+        version_id=new_product_version_db.version_id,
+        org_id=new_product_db.org_id,
+        name=new_product_version_db.name,
+        price=new_product_version_db.price,
+        measurement_unit=new_product_version_db.measurement_unit,
+    )
+
+
+def get_product(product_id: UUID, timestamp: Optional[datetime]) -> ProductSchema:
+    session = db_connector.get_session()
+    try:
+        product = session.query(Product).filter(Product.id == product_id).one()
+
+        if product.is_deleted:
+            raise NotFoundException(message="Product is deleted")
+
+        if timestamp:
+            product_version = (
+                session.query(ProductVersion)
+                .filter(ProductVersion.valid_from <= timestamp)
+                .filter(ProductVersion.valid_to >= timestamp)
+                .one()
+            )
+        else:
+            product_version = (
+                session.query(ProductVersion)
+                .filter(ProductVersion.product_id == product_id)
+                .filter(ProductVersion.valid_to == "9999-12-31 23:59:59.999999")
+                .one()
+            )
+
+    except NoResultFound:
+        raise NotFoundException(message="Product Details not found")
+
+    return ProductSchema(
+        product_id=product.id,
+        reorder_level=product.reorder_level,
+        version_id=product_version.version_id,
+        org_id=product.org_id,
+        name=product_version.name,
+        price=product_version.price,
+        measurement_unit=product_version.measurement_unit,
+    )
+
+
+def update(product_id: UUID, product_data: ProductUpdate) -> ProductSchema:
+    session = db_connector.get_session()
+    try:
+        existing_product = session.query(Product).filter(Product.id == product_id).one()
+
+        latest_product_version = (
+            session.query(ProductVersion)
+            .filter(ProductVersion.product_id == product_id)
+            .filter(ProductVersion.valid_to == "9999-12-31 23:59:59.999999")
+            .one()
+        )
+    except NoResultFound:
+        raise NotFoundException(message="Product Details not found")
+
+    current_time = datetime.now()
+    latest_product_version.valid_to = current_time
+
+    updated_product_version_data = {
+        "name": product_data.name or latest_product_version.name,
+        "price": product_data.price or latest_product_version.price,
+        "measurement_unit": product_data.measurement_unit or latest_product_version.measurement_unit,
+    }
+    version_id = latest_product_version.version_id
+
+    if not (
```

Comment:
Remember boolean algebra?

Can be re-written as
```py
if (
        updated_product_version_data["name"] != latest_product_version.name
        or updated_product_version_data["price"] != latest_product_version.price
        or updated_product_version_data["measurement_unit"] != latest_product_version.measurement_unit
    ):
```

Is more efficient as it matches fast rather than checking each condition.

Imo, the best solution for this would be a generator based evaluation with `any`
Assuming both are dicts
`if any(a[k] != b[k] for k in ["name", "measurement", "price"]):`